src/nlsolve/newton.jl

```diff
@@ -104,30 +104,51 @@ end
   @unpack z,tmp,ztmp,γ,iter,cache = nlsolver
   @unpack W_γdt,ustep,tstep,k,atmp,dz,W,new_W,invγdt,linsolve,weight = cache
 
-  mass_matrix = integrator.f.mass_matrix
   f = nlsolve_f(integrator)
+  isdae = f isa DAEFunction
+
+  if !isdae
+    mass_matrix = integrator.f.mass_matrix
+  end
 
-  @.. ustep = tmp + γ * z
-  f(k, ustep, p, tstep)
+  if isdae
+    @.. ustep = tmp + γ * z * inv(dt)
+    @.. ztmp = uprev + z
+    f(k, ustep, ztmp, p, tstep)
+  else
+    @.. ustep = tmp + γ * z
+    f(k, ustep, p, tstep)
+  end
   if DiffEqBase.has_destats(integrator)
     integrator.destats.nf += 1
   end
 
-  if mass_matrix === I
-    @.. ztmp = (dt * k - z) * invγdt
+  if isdae
+    b = vec(k)
   else
-    mul!(vec(ztmp), mass_matrix, vec(z))
-    @.. ztmp = (dt * k - ztmp) * invγdt
+    if mass_matrix === I
+      @.. ztmp = (dt * k - z) * invγdt
+    else
+      mul!(vec(ztmp), mass_matrix, vec(z))
+      @.. ztmp = (dt * k - ztmp) * invγdt
+    end
+    b = vec(ztmp)
   end
 
-  # update W
-  if W isa DiffEqBase.AbstractDiffEqLinearOperator
-    update_coefficients!(W, ustep, p, tstep)
+  if isdae
```

Comment:
This is a mess, but should get replaced by the `jvp` mechanism anyways so I'll let it slide.